docs/generic-templates.md

```diff
@@ -0,0 +1,173 @@
+
+class (Creatable t, Choice t c r) => Exercisable t c r | t c -> r where
+  exercise : ContractId t -> c -> Update r
+```
+
+## Desugaring generic template definitions
+
+Desugarting the generic template definition of `T` from above will generate the following data definitions type class instances:
+```haskell
+data T a = T with f_1 : U_1; ...
+
+instance D a => Template (T a)
+  signatory this@T{..} = sign
+
+
+data C a = C with g_1 : V_1; ...
+
+instance D a => Choice (T a) (C a) R where
+  controller this@T{..} arg@C{..} = ctrl
+  action self this@T{..} arg@C{..} = act
+```
+Extending this desugaring to more than one type parameter is trivial.
+
+## Desugaring generic template instantiations
+
+Desugaring the generic template instantion `I = T S` from above yields:
+```haskell
+newtype I = MkI with unI : T S
+
+instance LfTemplate I where
+  lfSignatory = signatory . unI
+
+instance Creatable (T S) where
+  create t = fmap coerceContractId . lfCreate @I . MkI
+  fetch = fmap unI . genFetch @I . coerceContractId
+
+
+instance LfChoice I (C S) R[S/a] where
+  lfController = controller . unI
+  lfAction cid = action (coerceContractId cid) . unI
+
+instance Exercisable (T S) (C S) R[S/a] where
+  exercise = lfExercise @I . coerceContractId
+```
+Extending this desugaring to more than one type parameter is trivial. Our actual implementation will most likely use auxiliary functions to simplify the method definition sof the generated instances.
+
+## Desugaring non-generic template definitions
+
+The template definition
+```haskell
+template T with
+    f_1 : U_1; ...
+  where
+    signatory sign
+
+    choice C : R with g_1 : V_1; ...
+      controller ctrl
+      do act
+```
+gets desugared into
+```haskell
+data T = T with f_1 : U_1; ...
+
+instance Template T where
+  signatory this@T{..} = sign
+
+instance LfTemplate T where
+  lfSignatory = signatory
+
+instance Creatable T where
+  create = lfCreate
+  fetch = lfFetch
+
+
+data C = C with g_1 : V_1; ...
+
+instance Choice T C R where
+  controller this@T{..} arg@C{..} = ctrl
+  action self this@T{..} arg@C{..} = act
+
+instance LfChoice T C R where
+  lfController = controller
+  lfAction = action
+
+instance Exercisable T C R where
+  exercise = lfExercise
+```
+
+## Risks and challenges
+
+The main challenge is generating the instances of `LfChoice` and `Exercisable` for the template instantiation `I = T S`. The only solution I can envision right now is to find all instances of the shape
+```haskell
+instance D a => Choice (T a) (C a) R where
+```
+in the module `T` is defined in. Unfortunately, this can't be done in the parser but needs to run after the renamer (to find the module `T` is defined in) and before the type checker. This compiler phase can reconstruct the original template instantiation `I = T S` easily if the parser produces the `newtype I`, `instance LfTemplate I` and `instance Creatable (T S)` part of the desugaring described above.
```

Comment:
I must admit, this sounds like it adds significant implementation complexity over what we've got going on today.

Reply:
I have an idea how to avoid this complexity. I'll play with it and iterate on the design.

Reply:
That's very good news. I hope it works out. Good luck!